pkg/plugins/gateway/algorithms/vtc/token_tracker.go

```diff
@@ -0,0 +1,176 @@
+}
+
+// InMemorySlidingWindowTokenTracker tracks tokens per user in a fixed-size sliding window (in-memory, thread-safe).
+type InMemorySlidingWindowTokenTracker struct {
+	mu          sync.RWMutex
+	windowSize  time.Duration
+	buckets     int
+	bucketUnit  TimeUnit
+	userBuckets map[string]map[int64]float64 // user -> windowStart -> tokenSum
+	config      *VTCConfig
+}
+
+// TokenTrackerOption is a function that configures a token tracker
+type TokenTrackerOption func(*InMemorySlidingWindowTokenTracker)
+
+func WithWindowSize(size int) TokenTrackerOption {
+	return func(t *InMemorySlidingWindowTokenTracker) {
+		t.buckets = size
+		t.windowSize = time.Duration(size) * timeUnitDuration[t.bucketUnit]
+	}
+}
+
+func WithTimeUnit(unit TimeUnit) TokenTrackerOption {
+	return func(t *InMemorySlidingWindowTokenTracker) {
+		t.bucketUnit = unit
+		t.windowSize = time.Duration(t.buckets) * timeUnitDuration[unit]
+	}
+}
+
+// TODO: add redis token tracker so that state is shared across plugin instances
+// NewInMemorySlidingWindowTokenTracker creates a new token tracker with configurable options
+func NewInMemorySlidingWindowTokenTracker(config *VTCConfig, opts ...TokenTrackerOption) *InMemorySlidingWindowTokenTracker {
+	// Default values
+	size := tokenTrackerWindowMinutes
+	if size < 1 {
+		size = defaultTokenTrackerWindowMinutes
+	}
+
+	tracker := &InMemorySlidingWindowTokenTracker{
+		windowSize:  time.Duration(size) * timeUnitDuration[Minutes],
+		buckets:     size,
+		bucketUnit:  Minutes,
+		userBuckets: make(map[string]map[int64]float64),
+		config:      config,
+	}
+
+	for _, opt := range opts {
+		opt(tracker)
+	}
+
+	return tracker
+}
+
+func (t *InMemorySlidingWindowTokenTracker) GetTokenCount(ctx context.Context, user string) (float64, error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if user == "" {
+		return 0, nil
+	}
+	buckets, ok := t.userBuckets[user]
+	if !ok {
+		return 0, nil
+	}
+
+	cutoffTime := time.Now().Add(-t.windowSize)
+	cutoff := t.bucketUnit.toTimestamp(cutoffTime)
+
+	sum := float64(0)
+	for ts, val := range buckets {
+		if ts >= cutoff {
+			sum += val
+		}
+	}
+
+	// Prune old buckets
+	for ts := range buckets {
+		if ts < cutoff {
+			delete(buckets, ts)
+		}
+	}
+	return sum, nil
+}
+
+func (t *InMemorySlidingWindowTokenTracker) UpdateTokenCount(ctx context.Context, user string, inputTokens, outputTokens float64) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if user == "" {
+		return nil
+	}
+	now := time.Now()
+
+	truncatedTime := t.bucketUnit.truncateTime(now)
+	windowStart := t.bucketUnit.toTimestamp(truncatedTime)
+
+	tokens := inputTokens*t.config.InputTokenWeight + outputTokens*t.config.OutputTokenWeight
+	buckets, ok := t.userBuckets[user]
+	if !ok {
+		buckets = make(map[int64]float64)
+		t.userBuckets[user] = buckets
+	}
+	buckets[windowStart] += tokens
+
+	cutoffTime := now.Add(-t.windowSize)
+	cutoff := t.bucketUnit.toTimestamp(cutoffTime)
+
+	// Prune old buckets
+	for ts := range buckets {
```

Comment:
`GetTokenCount` prunes the buckets as well. I suggest to extract the common pieces to separate method

Reply:
done